README.md
# Serve Static
A simple plugin to searve wordpress content as static site

## Development Details 
- Contributors: Mustafa Kamal Hossain
- Tags: Serverless databaseless, nosql, queryless, htmlSite
- Requires at least: 5.5
- Tested up to: 5.9
- Stable tag: 1.0.0
- Multisite Compitable: Yes
- License: GPLv3
- License URI: https://www.gnu.org/licenses/gpl-3.0.html

Migrate content from one live site to another.


## Description 

This plugin specificly build to make a wordpress site static, so that there will be no database and no queries at all. this will help to speed up the site. 

**Installation** 

This process defines you the steps to follow either you are installing through WordPress or Manually from FTP.

**From within WordPress**

1. Visit 'Plugins > Add New'
2. Search for Serve Static
3. Activate Serve Static from your Plugins page.
4. Go to "after activation" below.

**Manually**

1. Upload the `serve-staticr` folder to the `/wp-content/plugins/` directory
2. Activate Serve Static through the 'Plugins' menu in WordPress
3. Go to "after activation" below.

**After activation**

1. Go to the Serve Static settings page.
2. Enable the function from select box.
3. Enter the path where static site will be stored. Make sure the path parent must be under same usere which running the server. (www-data)
4. Enter subfolder name as the site placed. In our case our main WP site is in `bd-admin` folder, where as we are storing content inside `bd` folder and also bd folder is a subfolder of our site. so we will write `bd/bd-admin`. so that while creating the static version of this site our function will replace all links `example.com/bd/bd-admin` to `example.com`. 
5. Enter Local folder name (if any). if you want to store static site from a subfolder yow need to enter subfolder address. But do not use `/` in end. like your site is `example.com` and you want to keep all static content in `bd` subfolder. you need to write `bd` here. but if your site link is `example.com/bd` you do not need to put anything here
6. Enter Local server address. this part is used to replace link with `homeurl` or easy to say the static content site link will be here.
7. Check the post types you want to show in your static site. 
8. Press GET POST TYPES LIST Button.

## Now your settings are complete. 
You can build the site in two ways, one is Ajax calling from backend another is php from server.
* For ajax calling:
1. once you have completed all update and checked in WP site. if you think you are ready to build the site now. click on "Build site" link. 
2. check if all required post types list are exist or not. if you need to update setting please go settings page and fix that again.
3. Once you confirm the settings are ok, click build site button and wait until build is complete and the report is showing `DONE`
4. onece it is done it will be updated to github repo.

* For PHP Build from server:
1. As usual please confirm all changes are done and well reflected on WP site, also the settings are ok from settings page then click on build site and update repo link.
please wait until it shows the report of github repo.

And it is done.

** Upcoming features:
1. add github repo from backend.
2. Create github repo from backend.
3. Create corn job to to done the task.


changelog.txt
 = 1.0.0 - Apr, 24 2021 =
  * First realese on github
= 2.0.0 Aug, 24, 2022
 * Build site and update github repo.
 * Added WP CLI support
 * User bash script to build and update repo.
 * Functionality updated.
 * Reduced Build time.

admin/assets/admin.js
"use strict";

(function($) {
  $(document).ready(function() {

    var failed = [];
    var passed =[];


    $('.get-all').change( function(e){
      if(this.checked) {
        $(this).closest('.fetch-data').find('.id-specific-section').hide();
        $(this).closest('.fetch-data').find('.id-specific-section ul input:checkbox').each(function (index, value){
          $(value).prop('checked', true);
        })
      }else{
        $(this).closest('.fetch-data').find('.id-specific-section').show();
        $(this).closest('.fetch-data').find('.id-specific-section ul input:checkbox').each(function (index, value){
          $(value).prop('checked', false);
        })
      }
    });

    /**
     * make Ajax call
     * @value POST ID
     */
    function makeAjaxCall(mainURL, filestructures, homeurl, subfoldername){
      $.ajax({
        type: 'POST',
        url: ajax.ajaxurl,
        data: {
          'action' : "build_page",
          'mainurl' : mainURL,
          'filestructures' : filestructures,
          'homeurl' : homeurl,
          'subfoldername' : subfoldername
        },
        success: function(response){
          // if(response === 'null'){
          //   failed.push(postID);
          // }else{
          //   passed.push(postID);
          //   // passed[postID] = response;
          // }
//let modResponse =  response.replaceAll('&', '&amp;').replaceAll('<','&lt;').replaceAll('>', '&gt;');
          $('.progress-report > ul').append('<li>' + response + '</li>');
          console.log("the response is : ", response);
          // $('.report .items-finished').append('<li>' + postID + ' ended copying as ' + response + ' </li>');
        },
        async:false
      });
    }
    // Make ajax requests

    $('#server_static_builder').submit(function (e){
      e.preventDefault();

      const allURLS =  $.parseJSON($('input[name="allurl"]').val());
      const filestructures = $('input[name="filestructures"]').val();
      const homeurl =  $('input[name="homeurl"]').val();
      const subfoldername =  $('input[name="subfoldername"]').val();
      const sleep = (ms) => {
        return new Promise((resolve) => setTimeout(resolve, ms));
      };
      const getNumFruit = (mainURL, filestructures, homeurl, subfoldername) => {
        return sleep(1000).then((v) => {
          // $('.report .items-started').append('<li>' + $(post).val() + ' started to copying</li>');
          makeAjaxCall(mainURL, filestructures, homeurl, subfoldername );
        });
      };
      const forLoop = async (_) => {
        // $(".report").show();
        // $('.report .status').text("Please do not close the browser tab, posts are started to copying");
        // $('.report .loading').show();
        for (let index = 0; index < allURLS.length; index++) {
          const mainURL = allURLS[index];
          await getNumFruit(mainURL, filestructures, homeurl, subfoldername);
          // const numFruit = await getNumFruit(post);
          // console.log(numFruit);
        }
        // $('.report .status-ends').text("All task is finished you can close the tab now");
        // $('.report .loading').hide();
        // $(".report").hide();
        // console.log("Passed: ", passed, "Failed: ", failed);
        $.ajax({
          type: 'POST',
          url: ajax.ajaxurl,
          data: {
            'action' : "updateGithubRepo",
          },
          success: function(response){
            // if(response === 'null'){
            //   failed.push(postID);
            // }else{
            //   passed.push(postID);
            //   // passed[postID] = response;
            // }
            console.log(response);
            // $('.report .items-finished').append('<li>' + postID + ' ended copying as ' + response + ' </li>');
          },
          async:false
        });
        $('form').append("<div class='sucess' style='color: green; font-size:'60px'; text-align:'center'; '>Done</div>");
      };
      forLoop();

    });
  })
})(jQuery);


admin/gui/settings.php
<?php
 $checkboxArray = [];
 $flag = false;
 if( isset($_POST['list']) && !empty($_POST['checkboxArray']) && !empty($_POST['static_path'])){
    update_option('subfolder-name', $_POST['subfolder-name'], '' );
    update_option('localfolder', $_POST['localfolder'], '' );
   if( isset($_POST['local-server-address']) && !empty( $_POST['local-server-address'])){
    update_option('local-server-address', $_POST['local-server-address'], '' );
  }else{
     update_option('local-server-address', '', '' );
  }
   update_option('static_path', $_POST['static_path'], ABSPATH );
   update_option('make_static', (int)$_POST['static_status'], false);
   foreach($_POST['checkboxArray'] as $key => $value){
     array_push($checkboxArray, $value);
    }
    update_option('static_post_types', implode(',', $checkboxArray), true);
    $flag = true;
 }else{
   $checkboxArray = explode(',', get_option( 'static_post_types', [] )) ;
   $flag = true;
 }

?>
<div class="wrap">
  <h1 class="wp-heading-inline">SERVE STATIC SETTINGS</h1>
  <br><br>
  <hr class="wp-header-end" />
</div>
<?php 
 
  $post_types = get_post_types( Array( 'public' => true ) );
  unset($post_types['revision']);
  unset($post_types['nav_menu_item']);
  unset($post_types['custom_css']);
  unset($post_types['customize_changeset']);
  unset($post_types['oembed_cache']);
  unset($post_types['user_request']);
  unset($post_types['attachment']);
  unset($post_types['wp_global_styles']);
  unset($post_types['wp_navigation']);
  if($post_types){ ?>
    <div class="form">
      <form action="#" name="list_form" method="post">
      <!--  -->
      <label for="status">Function Status </label>
      <select name="static_status" id="static_status">
        <option value="1" <?= (get_option('make_static', false) )? 'selected':''; ?> >Enable</option>
        <option value="0"<?= (get_option('make_static', false) )? '':'selected'; ?>>Disable</option>
      </select>
      <br><br>
      <label for="static_path">Path to store data</label>
      <input type="text" name="static_path" id="static_path" value="<?= get_option('static_path',  ABSPATH );?>" size="200" />
      <br><br> 
      <label for="subfolder-name">Subfolder Name</label>
      <input type="text" name="subfolder-name" id="subfolder-name" value="<?= get_option('subfolder-name',  '' );?>" width="200" />
      <br><br> 
      <label for="localfolder">Local Folder</label>
      <input type="text" name="localfolder" id="localfolder" value="<?= get_option('localfolder',  'dist' );?>" width="200" />
      <br><p>Do not use slash in the end</p>
      <br><br> 
      <label for="local-server-address">Local Server Address</label>
      <input type="text" name="local-server-address" id="local-server-address" value="<?= get_option('local-server-address', home_url(). 'dist') ;?>" size="200" />
      <br><br> 
        <?php foreach($post_types as $key => $post_type){ ?>
          <label> 
            <input type="checkbox" name="checkboxArray[]" class="postTypeInputBox" value="<?= $post_type; ?>" <?= ( $flag && in_array($post_type , $checkboxArray)) ? 'checked' :''; ?>>
            <?= ucfirst($post_type); ?>
          </label><br>  
        <?php } ?>
        <br><br>
        <input type="submit" value="Get Post Types List" name="list" >
      </form>
    <br>
    <br>
    </div>
    <?php
  }
?>


admin/gui/build_site.php
<?php 
  /* Check is diirectory exist otherwise create your own directories */
  
  $filestructures =['root'=> get_option('static_path',  ABSPATH ), 'asset' => 'img', 'styles'=> 'css', 'js' => 'js', 'localfolder' => get_option( 'localfolder', 'dist' )];
  $localFolderPath = explode('/', $filestructures['localfolder']);
  $root = $filestructures['root'];
  foreach( $localFolderPath as $id => $folder){
    if (file_exists($root . $folder)) {
      echo "The directory ".$folder . " exists.<br>";
    } else {
      mkdir($root .  $folder, 0777);
      echo "The directory ".$folder . " was successfully created.<br>";
    }
    $root = $root . $folder . '/';
  }
  $subFolderName = get_option('subfolder-name',  '' );
  $allUrls = [];
  $homeURL = home_url();
  $post_types = get_option( 'static_post_types', '' );
  $postTypes = explode(',', $post_types);
  $posts = get_posts([
        'posts_per_page' => -1,
        'post_status' => 'publish',
        'post_type' => $postTypes,
      ]);
  foreach($posts as $key => $post){
   array_push($allUrls , get_permalink($post) );
  }
  $terms = get_terms('ml-expertinsights-category', ['hide_empty' => true]);
  foreach($terms as $key => $value){
    array_push($allUrls , get_category_link($value) );
  }
?>

<div class="wrap">
  <h1 class="wp-heading-inline">Build Static Site</h1>
  <br><br>
  <hr class="wp-header-end">

  <div class="form">
  <div class="progress-report">
  <ul></ul>
  </div>
    <form action="#" id="server_static_builder" name="server_static_builder" method="post">
      <input type="hidden" name="allurl" id="allurl" value='<?= json_encode($allUrls);?>'>
      <input type="text" name="localfolder" id="localfolder" value='<?= get_option('localfolder',  'dist' );?>'>
      <input type="hidden" name="homeurl" id="homeurl" value="<?= $homeURL;?>">
      <input type="hidden" name="subfoldername" id="subfoldername" value="<?= $subFolderName;?>">
      <input type="text" name="post_types" id="post_types" value="<?= $post_types; ?>" size="150">
      <input type="submit" value="Build" name="submit" style="position: fixed; bottom: 50px; right: 2%;">
    </form>

  </div>
</div>


serve-static.php
<?php
/**
 * @link    http://mkhossain.com/development/plugins/serve-static
 * @package Serve Static
 * @since   1.0.0
 * @version 2.0.0
 *
 * @wordpress-plugin
 * Plugin Name: Serve Static
 * Plugin URI: http://mkhossain.com/development/plugins/serve-static
 * Description: This plugin will help to make a static site using WordPress, in this way ther will be no server required at all, no RDS or Database. 
 * Author: MD Mustafa Kamal Hossain
 * Version: 2.0.0
 * Author URI: http://mkhossain.com
 * Text Domain: serve_static
 * Domain Path: /languages
 */


 // Make sure we don't expose any info if called directly
if ( ! defined( 'ABSPATH' ) ) {
  echo 'Hi there!  I\'m just a plugin, not much I can do when called directly.';
  exit;
}


// If this file is called directly, abort.
if (!defined('WPINC')) {
  die;
}

if ( ! defined( 'SERVE_STATIC_FILE' ) ) {
  define( 'SERVE_STATIC_FILE', __FILE__ );
}

if ( ! defined( 'SERVE_STATIC_PATH' ) ) {
  define( 'SERVE_STATIC_PATH', plugin_dir_path(SERVE_STATIC_FILE ));
}

/**
 * Currently plugin version.
 * Start at version 1.0.0 and use SemVer - https://semver.org
 */
define( 'SERVE_STATIC_VERSION', '2.0.0' );

/**
 * The code that runs during plugin activation.
 * This action is documented in includes/activator.php
 */
function activate_serve_static() {
	require_once plugin_dir_path( __FILE__ ) . 'includes/activator.php';
}

/**
 * The code that runs during plugin deactivation.
 * This action is documented in includes/deactivator.php
 */
function deactivate_serve_static() {
	require_once plugin_dir_path( __FILE__ ) . 'includes/deactivator.php';
}

register_activation_hook( __FILE__, 'activate_serve_static' );
register_deactivation_hook( __FILE__, 'deactivate_serve_static' );


/**
 * Register a custom menu page.
 */
function serve_static_add_settings_page() {
  add_menu_page(
      __( 'Serve Static', 'serve_static' ),
      'Serve Static Settings',
      'manage_options',
      'sarve-static/options-settings.php',
      'nss_add_settings_page',
      plugins_url( 'serve-static/images/icon.png' ),
      6
  );
  $submenu = [];
  $submenu[] = add_submenu_page(
    'sarve-static/options-settings.php',
    __( 'Build site Only', 'serve_static' ),
    __( 'Build site Only', 'serve_static' ),
    'manage_options',
    'serve-static/build-site.php',
    'nss_build_site',
    6
  );
  $submenu[] = add_submenu_page(
    'sarve-static/options-settings.php',
    __( 'Build Site & Update Repo', 'serve_static' ),
    __( 'Build Site & Update Repo', 'serve_static' ),
    'manage_options',
    'serve-static/update-repo.php',
    'nss_update_repo',
    6
  );
}
add_action( 'admin_menu', 'serve_static_add_settings_page' );


function nss_add_settings_page(){
require_once('admin/gui/settings.php');
}


function nss_build_site(){
require_once('admin/gui/build_site.php');
}


function nss_update_repo(){
require_once('admin/gui/update_repo.php');
}


/**
 * Enqueue a script in the WordPress admin on copier page.
 *
 * @param int $hook Hook suffix for the current admin page.
 */
function nss_add_custom_css_to_admin( $hook ) {
  if ( 'toplevel_page_sarve-static/options-settings' == $hook || 'serve-static-settings_page_serve-static/build-site' == $hook ) {
    
    wp_enqueue_script( 'server_static_custom_js', plugin_dir_url( __FILE__ ) . 'admin/assets/admin.js', array('jquery'), '1.0' );
    wp_localize_script(
      'server_static_custom_js',
      'ajax',
      array(
          'ajaxurl' => admin_url('admin-ajax.php')
      )
    );
    wp_enqueue_style( 'server_static_custom_css', plugin_dir_url( __FILE__ ) . 'admin/assets/admin.css', '', '1');
  }
}
add_action( 'admin_enqueue_scripts', 'nss_add_custom_css_to_admin' );


function nss_build_page($url = NULL , $home = NULL ){
  $localServer = get_option('local-server-address', '');
  $cliFlag = 0;
  $subfoldername = get_option('subfolder-name',  '' );
  if($url == NULL || $home == NULL ){
    $mainURL = $_POST['mainurl'];
    $homeURL = $_POST['homeurl'];
  }else{
    $mainURL = $url;
    $homeURL = $home;
    $cliFlag = 1;
  }

  $pageFlag = 0;
  $filestructures =['root'=> get_option('static_path',  ABSPATH ), 'asset' => 'img', 'styles'=> 'css', 'js' => 'js', 'localfolder' => get_option( 'localfolder', 'dist')]; 
  $html = file_get_contents( $mainURL);

  $stringToReplace = [
    '<link rel="stylesheet" href="https://dev-test.monstar-lab.com/bd/bd-admin/assets/themes/monstar_lab/style.css">',
    '<link rel="icon" type="image/ico" href="https://dev-test.monstar-lab.com/bd/bd-admin/assets/themes/monstar_lab/public/images/favicon.ico">',
    '<link rel="icon" type="image/png" href="https://dev-test.monstar-lab.com/bd/bd-admin/assets/themes/monstar_lab/public/images/favicon.png">',
    '<link rel="apple-touch-icon" href="https://dev-test.monstar-lab.com/bd/bd-admin/assets/themes/monstar_lab/public/images/apple-touch-icon.png">',
    '<link rel="https://api.w.org/" href="https://dev-test.monstar-lab.com/bd/bd-admin/wp-json/" />',
    '<link rel="alternate" type="application/json" href="https://dev-test.monstar-lab.com/bd/bd-admin/wp-json/wp/v2/pages/2151" />',
    '<link rel="canonical" href="https://dev-test.monstar-lab.com/bd/bd-admin/" />',
    '<link rel=\'shortlink\' href=\'https://dev-test.monstar-lab.com/bd/bd-admin/\' />',
    '<link rel="alternate" type="application/json+oembed" href="https://dev-test.monstar-lab.com/bd/bd-admin/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fdev-test.monstar-lab.com%2Fbd%2Fbd-admin%2F" />',
    '<link rel="alternate" type="text/xml+oembed" href="https://dev-test.monstar-lab.com/bd/bd-admin/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fdev-test.monstar-lab.com%2Fbd%2Fbd-admin%2F&#038;format=xml" />',
    '<script type=\'text/javascript\' src=\'https://dev-test.monstar-lab.com/bd/bd-admin/wp-includes/js/jquery/jquery.min.js\' id=\'jquery-core-js\'></script>',
    "<script type='text/javascript' src='https://dev-test.monstar-lab.com/bd/bd-admin/wp-includes/js/jquery/jquery-migrate.min.js' id='jquery-migrate-js'></script>",
    '/(<style id=\'global-styles-inline-css\' type=\'text\/css\'>)*(^<\/style>)/m'
  ];

  $stringToReplaceWith = [
    '',
    '',
    '<link rel="icon" type="image/png" href="https://dev-bd-website.s3.ap-northeast-1.amazonaws.com/bd/bd-admin/assets/uploads/2022/07/favicon.png">',
    '<link rel="apple-touch-icon" href="https://dev-bd-website.s3.ap-northeast-1.amazonaws.com/bd/bd-admin/assets/uploads/2022/07/apple-touch-icon.png">',
    '',
    '',
    '',
    '',
    '',
    '',
    '<script src="https://code.jquery.com/jquery-3.6.0.slim.min.js" integrity="sha256-u7e5khyithlIdTpu22PHhENmPcRdFiHRjhAuHcs05RI=" crossorigin="anonymous"></script>',
    '<script type=\'text/javascript\' src=\'https://cdnjs.cloudflare.com/ajax/libs/jquery-migrate/3.4.0/jquery-migrate.min.js\' ></script>',
    ''
  ];

  $html = str_replace($stringToReplace, $stringToReplaceWith, $html);

  /* Firstly make the folder to write a file */
  $url = substr_replace(substr(parse_url($mainURL)['path'], 1), "", -1) ;
  $subURL = str_replace($homeURL,"", $url);
  $subURL2 = str_replace($subfoldername ."/","", $subURL);
  if($subURL2 === $subfoldername){
    $subURL2 = '';
  }
  $paths = explode('/', $subURL2);
  $localwslash = ($filestructures['localfolder'] == '')? '' : $filestructures['localfolder'] . '/';
  $thisPath = $filestructures['root'] . $localwslash;
  foreach($paths as $key => $path ){
    $filename = $thisPath . $path . '/';
    if (!file_exists($filename)) {
      mkdir( $thisPath .  $path, 0777);
//      echo "The directory ". $path . " was successfully created.<br>";
    } else {
//      echo "The directory ". $path . " exists.<br>";
    }
    $thisPath = $thisPath . $path . '/';
  }

  $re = '/(<a\s[^>]*href\s*=\s*([\"\']?))*(?:\/page\/)([^\" >]*?)/Ui';
  preg_match_all($re, $html, $matches, PREG_SET_ORDER, 0);
  if( !empty($matches)){
    $pageFlag = 1;
    $matches = end($matches);
    $prev_page =  $matches[0];
    $pageURL = rtrim( preg_replace('/\/page\/\d+/', '', $mainURL, 1)  , '/') . $prev_page ;
  }else{
    $pageFlag = 0;
  }

  // $pattern = '/(<a\s[^>]*href\s*=\s*([\"\']?))(?:https:\/\/dev\-test.monstar\-lab.com\/bd)([^\" >]*?)/Ui';
  $rehome = str_replace("/","\/",$homeURL);
  $rehome = str_replace("-","\-",$rehome);
  $pattern = '/(<a\s[^>]*href\s*=\s*([\"\']?))(?:' . $rehome . ')([^\" >]*?)/Ui';
  // $html = preg_replace($pattern, '${1}http://cache-bd.localhost${3}', $html);
  $html = preg_replace($pattern, '${1}'. $localServer .'${3}', $html);
/* Now write html in file */

  $myfile = fopen($thisPath . "index.html", "w") or die("Unable to open file!");
  fwrite($myfile, $html);
  fclose($myfile);
  //echo $thisPath .'index.html' ;

  if($pageFlag == 1){
    nss_build_paged_metarial( $pageURL, $prev_page, $homeURL , $subfoldername, $cliFlag );
  }
  if(! $cliFlag) exit();
}

add_action( 'wp_ajax_build_page', 'nss_build_page' );
add_action( 'wp_ajax_nopriv_build_page', 'nss_build_page' );


function nss_build_paged_metarial($url, $prev_page = 0, $homeURL , $subfoldername, $cliFlag ){
  $localServer = get_option('local-server-address', '');
  $mainURL = $url;
  $pageFlag = 0;
  $filestructures =['root'=> get_option('static_path',  ABSPATH ), 'asset' => 'img', 'styles'=> 'css', 'js' => 'js', 'localfolder' => get_option( 'localfolder', 'dist')]; 
  $html = file_get_contents( $mainURL);
  /* Firstly make the folder to write a file */

  $stringToReplace = [
    '<link rel="stylesheet" href="https://dev-test.monstar-lab.com/bd/bd-admin/assets/themes/monstar_lab/style.css">',
    '<link rel="icon" type="image/ico" href="https://dev-test.monstar-lab.com/bd/bd-admin/assets/themes/monstar_lab/public/images/favicon.ico">',
    '<link rel="icon" type="image/png" href="https://dev-test.monstar-lab.com/bd/bd-admin/assets/themes/monstar_lab/public/images/favicon.png">',
    '<link rel="apple-touch-icon" href="https://dev-test.monstar-lab.com/bd/bd-admin/assets/themes/monstar_lab/public/images/apple-touch-icon.png">',
    '<link rel="https://api.w.org/" href="https://dev-test.monstar-lab.com/bd/bd-admin/wp-json/" />',
    '<link rel="alternate" type="application/json" href="https://dev-test.monstar-lab.com/bd/bd-admin/wp-json/wp/v2/pages/2151" />',
    '<link rel="canonical" href="https://dev-test.monstar-lab.com/bd/bd-admin/" />',
    '<link rel=\'shortlink\' href=\'https://dev-test.monstar-lab.com/bd/bd-admin/\' />',
    '<link rel="alternate" type="application/json+oembed" href="https://dev-test.monstar-lab.com/bd/bd-admin/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fdev-test.monstar-lab.com%2Fbd%2Fbd-admin%2F" />',
    '<link rel="alternate" type="text/xml+oembed" href="https://dev-test.monstar-lab.com/bd/bd-admin/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fdev-test.monstar-lab.com%2Fbd%2Fbd-admin%2F&#038;format=xml" />',
    '<script type=\'text/javascript\' src=\'https://dev-test.monstar-lab.com/bd/bd-admin/wp-includes/js/jquery/jquery.min.js\' id=\'jquery-core-js\'></script>',
    "<script type='text/javascript' src='https://dev-test.monstar-lab.com/bd/bd-admin/wp-includes/js/jquery/jquery-migrate.min.js' id='jquery-migrate-js'></script>",
    '/(<style id=\'global-styles-inline-css\' type=\'text\/css\'>)*(^<\/style>)/m'
  ];

  $stringToReplaceWith = [
    '',
    '',
    '<link rel="icon" type="image/png" href="https://dev-bd-website.s3.ap-northeast-1.amazonaws.com/bd/bd-admin/assets/uploads/2022/07/favicon.png">',
    '<link rel="apple-touch-icon" href="https://dev-bd-website.s3.ap-northeast-1.amazonaws.com/bd/bd-admin/assets/uploads/2022/07/apple-touch-icon.png">',
    '',
    '',
    '',
    '',
    '',
    '',
    '<script src="https://code.jquery.com/jquery-3.6.0.slim.min.js" integrity="sha256-u7e5khyithlIdTpu22PHhENmPcRdFiHRjhAuHcs05RI=" crossorigin="anonymous"></script>',
    '<script type=\'text/javascript\' src=\'https://cdnjs.cloudflare.com/ajax/libs/jquery-migrate/3.4.0/jquery-migrate.min.js\' ></script>',
    ''
  ];

  $html = str_replace($stringToReplace, $stringToReplaceWith, $html);

  $url = substr_replace(substr(parse_url($mainURL)['path'], 1), "", -1) ;
  $subURL = str_replace($homeURL,"", $url);
  $subURL2 = str_replace($subfoldername ."/","", $subURL);
  $paths = explode('/', $subURL2);
  $localwslash = ($filestructures['localfolder'] == '')? '' : $filestructures['localfolder'] . '/';
  $thisPath = $filestructures['root'] . $localwslash;

  foreach($paths as $key => $path ){
    $filename = $thisPath . $path . '/';
    if (!file_exists($filename)) {
      mkdir( $thisPath .  $path, 0777);
//      echo "The directory ". $path . " was successfully created.<br>";
    } else {
 //     echo "The directory ". $path . " exists.<br>";
    }
    $thisPath = $thisPath . $path . '/';
  }
  $re = '/(<a\s[^>]*href\s*=\s*([\"\']?))*(?:\/page\/)([^\" >]*?)/Ui';
  preg_match_all($re, $html, $matches, PREG_SET_ORDER, 0);
  if( !empty($matches)){
    $matches = end($matches);
    if($matches[0] !== $prev_page) {
      $pageFlag = 1;
      $pageURL = preg_replace('/\/page\/\d+/', '', $mainURL, 1) . $matches[0] ;
      $prev_page = $matches[0];
    }else{
      $pageFlag = 0;
    }
  }else{
    $pageFlag = 0;
  }
  /* Now grab the content from url */
  // $pattern = '/(<a\s[^>]*href\s*=\s*([\"\']?))(?:https:\/\/dev\-test.monstar\-lab.com\/bd)([^\" >]*?)/Ui';
  $rehome = str_replace("/","\/",$homeURL);
  $rehome = str_replace("-","\-",$rehome);
  $pattern = '/(<a\s[^>]*href\s*=\s*([\"\']?))(?:' . $rehome . ')([^\" >]*?)/Ui';

  // $html = preg_replace($pattern, '${1}http://cache-bd.localhost${3}', $html);
  $html = preg_replace($pattern, '${1}' . $localServer . '${3}', $html);
  /* Now write html in file */
  $myfile = fopen($thisPath . "index.html", "w") or die("Unable to open file!");
  fwrite($myfile, $html);
  fclose($myfile);
//  echo $thisPath .'index.html' ;
  if($pageFlag  == 1){
    nss_build_paged_metarial( $pageURL, $prev_page, $homeURL , $subfoldername, $cliFlag );
  }
  if(! $cliFlag) exit();
}



function updateGithubRepo(){
  $a='/home/ubuntu';
  chdir($a);
  $output = shell_exec("sh static-process.sh 2>&1");
  echo "<pre>$output</pre>";
  return json_encode($output);
}


add_action( 'wp_ajax_updateGithubRepo', 'updateGithubRepo' );
add_action( 'wp_ajax_nopriv_updateGithubRepo', 'updateGithubRepo' );



/*  DISABLE GUTENBERG STYLE IN HEADER| WordPress 5.9 */
function nss_deregister_styles() {
  wp_dequeue_style( 'global-styles' );
}
add_action( 'wp_enqueue_scripts', 'nss_deregister_styles', 100 );


class SERVESTATIC_CLI {

	/**
	 * Returns 'Build Static Site'
	 *
	 * @since  1.0.0
	 * @author Mustafa Kamal Hossain
	 */
	public function build() {
		WP_CLI::line( 'Starting Genaretion of static content' );
    $homeURL = home_url();
    $post_types = get_option( 'static_post_types', '' );
    $postTypes = explode(',', $post_types);
    $posts = get_posts([
        'posts_per_page' => -1,
        'post_status' => 'publish',
        'post_type' => $postTypes,
    ]);
    
    $filestructures =['root'=> get_option('static_path',  ABSPATH ), 'asset' => '-img', 'styles'=> 'css', 'js' => 'js', 'localfolder' => get_option( 'localfolder', 'dist' )];
    $localFolderPath = explode('/', $filestructures['localfolder']);
    $root = $filestructures['root'];
    foreach( $localFolderPath as $id => $folder){
      if (file_exists($root . $folder)) {
        echo "The directory ".$folder . " exists.<br>";
      } else {
        WP_CLI::line($folder);
        mkdir($root .  $folder, 0777);
        echo "The directory ".$folder . " was successfully created.<br>";
      }
      $root = $root . $folder . '/';
    }
    $assoc_args = [$homeURL];
    $this->generate_posts_progress_bar($posts, $assoc_args);
	}

  /**
 * Displays progress bar to demonstrate progression through a time consuming process.
 *
 * @param Array $args Arguments in array format.
 * @param Array $assoc_args Key value arguments stored in associated array format.
 * @since 1.0.0
 * @author Mustafa Kamal Hossain
 */
public function generate_posts_progress_bar( $args= [], $assoc_args = [] ) {
  $desired_posts_to_generate = count($args);
  $progress = \WP_CLI\Utils\make_progress_bar( 'Generating Content', $desired_posts_to_generate );
  foreach( $args as $key=> $value){
    $link = get_permalink($value);
 //   WP_CLI::line( $value->post_title );
    nss_build_page($link, $assoc_args[0]);
    $progress->tick();
  }

  $progress->finish();
  WP_CLI::success("Content Genaration has Completed, Thanks for being Patients");
  updateGithubRepo();
  }

}

/**
 * Registers our command when cli get's initialized.
 *
 * @since  1.0.0
 * @author Mustafa Kamal
 */
function static_cli_register_commands() {
	WP_CLI::add_command( 'static', 'SERVESTATIC_CLI' );
}

add_action( 'cli_init', 'static_cli_register_commands' );


function make_static_using_CLI($post_id, $post, $update)  {
  //static-build.sh
  // $a='/home/ubuntu';
  // chdir($a);
  // $output = [];
  // $output["shell"] = shell_exec("sh static-build.sh 2>&1");
  // $output["exec"] = exec( "sudo wp static build --allow-root");
  // return json_encode($output);

  $homeURL = home_url();
  $post_types = get_option( 'static_post_types', '' );
  $postTypes = explode(',', $post_types);
  $posts = get_posts([
      'posts_per_page' => -1,
      'post_status' => 'publish',
      'post_type' => $postTypes,
  ]);
  
  $filestructures =['root'=> get_option('static_path',  ABSPATH ), 'asset' => 'img', 'styles'=> 'css', 'js' => 'js', 'localfolder' => get_option( 'localfolder', 'dist' )];
  $localFolderPath = explode('/', $filestructures['localfolder']);
  $root = $filestructures['root'];
  foreach( $localFolderPath as $id => $folder){
    if (file_exists($root . $folder)) {
    } else {
      mkdir($root .  $folder, 0777);
    }
    $root = $root . $folder . '/';
  }
  
  foreach( $posts as $key=> $value){
    $link = get_permalink($value);
    nss_build_page($link,$homeURL);
  }

  $terms = get_terms('ml-expertinsights-category', ['hide_empty' => true]);
  foreach($terms as $key => $value){
    nss_build_page( get_category_link($value) ,$homeURL);
  }
  updateGithubRepo();
}

//add_action('save_post', 'make_static_using_CLI', 30,3);


function make_static_using_CLI2()  {
  $a='/var/www/html/bd/bd-admin';
  chdir($a);
  $output = [];
  //$output["shell"] = shell_exec("sh static-build.sh 2>&1");
  $output["shell"] = shell_exec("wp static build");
  //$output["exec"] = exec( "sudo wp static build --allow-root");
  //updateGithubRepo();
  return json_encode($output);
}
